src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,142 @@
+            return None
+
+        svc = svclist[0]
+        self.log.info("newsvc:{}".format(svc))
+        self.log.info("newsvc.spec:{}".format(svc.spec))
+        self.log.info("new count:{}".format(abscount))
+
+        hosts = []
+        for h in svc.spec.placement.hosts:
+            hosts.append(h.hostname)
+        ps = PlacementSpec(hosts=hosts, count=abscount)
+        newsvc = ServiceSpec(service_type=svc.spec.service_type,
+                             service_id=svc.spec.service_id,
+                             placement=ps)
+        return newsvc
+
+    def spawn_mds(self, fs_name, count):
+        newsvc = self.update_daemon_count(fs_name, count+1)
+        if newsvc is None:
+            return
+
+        completion = self.apply_mds(newsvc)
+        self._orchestrator_wait([completion])
+        orchestrator.raise_if_exception(completion)
+
+    def kill_mds(self, fs_name, count):
+        newsvc = self.update_daemon_count(fs_name, count-1)
+        if newsvc is None:
+            return
+
+        completion = self.apply_mds(newsvc)
+        self._orchestrator_wait([completion])
+        orchestrator.raise_if_exception(completion)
+
+    def get_required_standby_count(self):
+        total = 1
+        assert self.fs_map is not None
+        for fs in self.fs_map.get('filesystems'):
+            self.log.info("getting standby_count_wanted")
+            total = max(total, fs.get('mdsmap').get('standby_count_wanted'))
+        return total
+
+    def get_current_standby_count(self):
+        # standbys are not grouped by filesystems in fs_map
+        # available = standby_replay + standby_active
+        assert self.fs_map is not None
+        return len(self.fs_map.get('standbys'))
+
+    def get_current_active_count(self, fs_name):
+        assert self.fs_map is not None
+        for fsys in self.fs_map.get('filesystems'):
+            if fsys.get('mdsmap').get('fs_name') == fs_name:
+                return fsys.get('mdsmap').get('max_mds')
+        return 0
+
+    def get_fs_name(self, index=0):
+        assert self.fs_map is not None
+        fs = self.fs_map.get('filesystems')[index]
+        self.log.info("fs:{}".format(fs))
+        fs_name = fs.get('mdsmap').get('fs_name')
+        self.log.info("fs_name:{}".format(fs_name))
+        return fs_name
+
+    def verify_and_manage_mds_instance(self):
+        standbys_required = self.get_required_standby_count()
+        standbys_current = self.get_current_standby_count()
+        standbys_active = self.get_current_active_count(self.get_fs_name())
+
+        self.log.info("standbys_required:{0}, standbys_current:{1}"
+                      .format(standbys_required, standbys_current))
+        total = standbys_current + standbys_active
+        if standbys_current > standbys_required:
+            # remove one mds at a time and wait for updated fs_map
+            self.log.info("killing standby mds ...")
+            self.kill_mds(self.get_fs_name(), total)
```

Comment:
Do you really need two methods `kill_mds()` and `spawn_mds()`, which are thin wrappers around `update_daemon_count()` method? Can't you just call `update_daemon_count()` from here.  You can add the following lines to `update_daemon_count()`
```   
        completion = self.apply_mds(newsvc)
        self._orchestrator_wait([completion])
        orchestrator.raise_if_exception(completion)
```
Then here,
```
if standbys_current > standbys_required:
       total = <expected mds count for the filesystem>
elif standbys_current < standbys_required:
      total = <some other expected mds count for the filesystem>

self.update_daemon_count(self.get_fs_name(), total)
```